sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/BeamSqlFnExecutor.java

```diff
@@ -171,36 +171,39 @@ static BeamSqlExpression buildExpression(RexNode rexNode) {
         //     node.getSqlTypeName() = DECIMAL (the actual internal storage format)
         // So we need to do a convert here.
         // check RexBuilder#makeLiteral for more information.
+
         SqlTypeName realType = node.getType().getSqlTypeName();
         Object realValue = value;
-        if (type == SqlTypeName.DECIMAL) {
-          BigDecimal rawValue = (BigDecimal) value;
+
+        if (SqlTypeName.NUMERIC_TYPES.contains(type)) {
           switch (realType) {
             case TINYINT:
-              realValue = (byte) rawValue.intValue();
+              realValue = SqlFunctions.toByte(value);
               break;
             case SMALLINT:
-              realValue = (short) rawValue.intValue();
+              realValue = SqlFunctions.toShort(value);
               break;
             case INTEGER:
-              realValue = rawValue.intValue();
+              realValue = SqlFunctions.toInt(value);
               break;
             case BIGINT:
-              realValue = rawValue.longValue();
+              realValue = SqlFunctions.toLong(value);
+              break;
+            case FLOAT:
+              realValue = SqlFunctions.toFloat(value);
+              break;
+            case DOUBLE:
+              realValue = SqlFunctions.toDouble(value);
               break;
             case DECIMAL:
-              realValue = rawValue;
+              realValue = SqlFunctions.toBigDecimal(value);
               break;
             default:
               throw new IllegalStateException(
-                  "type/realType mismatch: " + type + " VS " + realType);
-          }
-        } else if (type == SqlTypeName.DOUBLE) {
-          Double rawValue = (Double) value;
-          if (realType == SqlTypeName.FLOAT) {
-            realValue = rawValue.floatValue();
+                  "type / realType mismatch: " + type + " <-> " + realType);
```

Comment:
nit: explain a bit more in the message? E.g. "Attempted convert node zzz of type xxx to yyy. This conversion is unsupported"

Reply:
Addressed. Update later.